### PR TITLE
Check operator CSV file with schema before processing

### PR DESCRIPTION
**Pull request by ben-alkov**
- Uses a reduced schema which only checks the parts of the CSV which we
    really care about
- The schema is embedded in the source as a Python Dict

Signed-off-by: Ben Alkov <ben.alkov@redhat.com>

CLOUDBLD-2830

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] JSON/YAML configuration changes are updated in the relevant schema
- [ ] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates
- [ ] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/utils/operator.py [hunk 1]
```diff
@@ -293,6 +302,180 @@ class OperatorCSV(object):
     # Annotation keys that are expected to contain pullspecs
     _known_annotation_keys = ("containerImage",)
 
+    _mini_csv_schema_json = '''
```
Comment (MartinBasti) [inline atomic_reactor/utils/operator.py:305]: Please define schema as a constant on module level

File: atomic_reactor/utils/operator.py [hunk 2]
```diff
@@ -293,6 +302,180 @@ class OperatorCSV(object):
...
+{
```
Comment (MartinBasti) [inline atomic_reactor/utils/operator.py:306]: Wold be nice to have py code in py file isntead of JSON

File: atomic_reactor/utils/operator.py [hunk 3]
```diff
@@ -304,14 +487,34 @@ def __init__(self, path, data, pullspec_heuristic=default_pullspec_heuristic):
...
+    def check_csv(cls, path, data):
         try:
             if data.get("kind") != OPERATOR_CSV_KIND:
                 raise NotOperatorCSV("Not a ClusterServiceVersion")
         except AttributeError:
             raise NotOperatorCSV("File does not contain a YAML object")
-        self.path = path
-        self.data = data
-        self._pullspec_heuristic = pullspec_heuristic
+        # Validate CSVs against schema before *any* other processing
```
Comment (MartinBasti) [inline atomic_reactor/utils/operator.py:502]: Please use common method for validation we use elsewhere, or refactor it as needed

**Review by MartinBasti — CHANGES_REQUESTED** [review]

File: atomic_reactor/utils/operator.py (hunk 3 above)
Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:502]: Which common method? `osbs-client:read_yaml()`? How will I be able to use that when it requires a path to a schema file, and I have a schema embedded in a Python dictionary?

File: atomic_reactor/utils/operator.py [hunk 4]
```diff
@@ -293,6 +302,180 @@ class OperatorCSV(object):
...
+  "type": "object",
+  "required": [
+    "metadata",
+    "spec"
+  ],
+  "properties": {
+    "metadata": {
+      "type": "object",
+      "properties": {
+        "annotations": {
```
[the remark this answers is not in the document]
Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:317]: OK... Do we want to include "annotations" in the schema everywhere they might possibly occur (source says "they may be found under any key")?

File: atomic_reactor/utils/operator.py [hunk 5]
```diff
@@ -293,6 +302,180 @@ class OperatorCSV(object):
...
+          }
+        },
+        "relatedImages": {
+          "description": "Related images",
+          "type": "array",
+          "items": {
+            "description": "A single related image",
+            "type": "object",
+            "required": [
+              "name",
```
[the remark this answers is not in the document]
Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:460]: Oops :|

File: atomic_reactor/utils/operator.py (hunk 1 above)
Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:305]: ACK

File: atomic_reactor/utils/operator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/utils/operator.py:317]: I'd just skip them for now, code for handling them looks robust enough

File: atomic_reactor/utils/operator.py (hunk 3 above)
Comment (MartinBasti) [reply atomic_reactor/utils/operator.py:502]: Split current read_yaml into two methods (I dunno why it was not done as that before as those things are orthogonal)


```python
def validate_with_schema(data_as_py_object, schema_as_dict):
    # validation code from current read_yaml, raisess exceptions if data are invalid
   ....
   pass

def read_yaml(....):
    # load files, first part of current read_yaml
    ....
    data = ...
    schema = ...
    validate_with_schema(data, schema)
    return data
```

Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:502]: ACK, opened https://github.com/containerbuildsystem/osbs-client/pull/1009

File: atomic_reactor/utils/operator.py [hunk 6]
```diff
@@ -162,6 +338,12 @@ class NotOperatorCSV(Exception):
     """
 
 
+class ValidationError(Exception):
+    """
+    CSV validation against schema failed
+    """
```
Comment (tkdchen) [inline atomic_reactor/utils/operator.py:344]: Where is this error raised?

And looks this error is specific to a failure CSV validation, rename it to a more specific name? `ValidationError` is too general.

File: atomic_reactor/utils/operator.py [hunk 7]
```diff
@@ -304,11 +486,7 @@ def __init__(self, path, data, pullspec_heuristic=default_pullspec_heuristic):
             implementation will be used. For more information, see the
             default_pullspec_heuristic() function in this module.
         """
-        try:
-            if data.get("kind") != OPERATOR_CSV_KIND:
-                raise NotOperatorCSV("Not a ClusterServiceVersion")
-        except AttributeError as exc:
-            raise NotOperatorCSV("File does not contain a YAML object") from exc
+        check_csv(data, _mini_csv_schema)
```
Comment (tkdchen) [inline atomic_reactor/utils/operator.py:489]: What's the opinion to separate these lines into a function?

The schema is defined in module level, is it necessary to pass into the function via a parameter? Just reference it directly in that function?

File: atomic_reactor/utils/operator.py [hunk 8]
```diff
@@ -23,6 +24,181 @@
 
 OPERATOR_CSV_KIND = "ClusterServiceVersion"
 
+# Schema derived from https://github.com/operator-framework/api/blob/e29d40c2eb3eab6a2789a671886b8966ce8e3c0a/crds/operators.coreos.com_clusterserviceversions.yaml#L40  # noqa E501
+_mini_csv_schema = {
```
Comment (tkdchen) [inline atomic_reactor/utils/operator.py:28]: Any reason not put this schema into the `schemas/`?

Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:28]: @MartinBasti? -^

File: atomic_reactor/utils/operator.py (hunk 7 above)
Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:489]: @tkdchen; You're abslutely right, it should probably be a class method. IIRC, my goal was to simplify `OperatorCSV.__init__()`.

File: atomic_reactor/utils/operator.py (hunk 6 above)
Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:344]: @tkdchen; L# 799 in operator.py

Good point about the name.

File: atomic_reactor/utils/operator.py (hunk 7 above)
Comment (ben-alkov) [reply atomic_reactor/utils/operator.py]: @tkdchen;
I'm not smart enough to unwind how to bypass schema checking in something like e.g. `utils/test_operator.py::test_get_pullspecs_some_locations()`.

Having `check_csv()` as a module function (it doesn't access cls or self) means I can easily mock it out of the unit tests.

Good catch on the arg :)

File: atomic_reactor/utils/operator.py (hunk 8 above)
Comment (MartinBasti) [reply atomic_reactor/utils/operator.py:28]: https://github.com/containerbuildsystem/operator-manifest  this code will be synced to shared library. Eventually, there will be only shared library for operator manipulations in future

**Review by MartinBasti — APPROVED** [review]

**Comment by MartinBasti** [comment]
Please fix or silent pep8 linter
```
./atomic_reactor/utils/operator.py:105:101: E501 line too long (101 > 100 characters)
4
./atomic_reactor/utils/operator.py:109:101: E501 line too long (103 > 100 characters)
5
./atomic_reactor/utils/operator.py:110:101: E501 line too long (104 > 100 characters)
6
./atomic_reactor/utils/operator.py:114:101: E501 line too long (103 > 100 characters)
7
./atomic_reactor/utils/operator.py:115:101: E501 line too long (104 > 100 characters)
```

**Comment by ben-alkov** [comment]
> Please fix or silent pep8 linter
> 
> ```
> ./atomic_reactor/utils/operator.py:105:101: E501 line too long (101 > 100 characters)
> ./atomic_reactor/utils/operator.py:109:101: E501 line too long (103 > 100 characters)
> ./atomic_reactor/utils/operator.py:110:101: E501 line too long (104 > 100 characters)
> ./atomic_reactor/utils/operator.py:114:101: E501 line too long (103 > 100 characters)
> ./atomic_reactor/utils/operator.py:115:101: E501 line too long (104 > 100 characters)
> ```

I'm not sure what to do about this, other than ignore it. Any formatting options I try result in *some* kind of flake8 complaint; line length/indentation...

This is strictly due to the embedded schema.

File: atomic_reactor/utils/operator.py (hunk 8 above)
Comment (ben-alkov) [reply atomic_reactor/utils/operator.py:28]: @tkdchen, @mkosiarc; tldr; Maintaining *one* file in common between this repo and [operator_manifest](https://github.com/containerbuildsystem/operator-manifest) is easier than maintaining more than one.

This is a temporary situation, i.e. until operator_manifest truly becomes its own thing.

**Review by mkosiarc — APPROVED** [review]

**Pull request merged**